### ISSUE TITLE
Support embedded version of the service

### ISSUE THREAD

**Issue by mfateev** [temporalio/temporal#298]
For small scale on prem deployments having ability to run Temporal as a library which has only DB as a dependency would be really great.

**Comment by yiminc** [comment]
This is currently possible to do using SQL (MySQL/PostgresSQL/SQLite) as both default and visibility store. Is that what is requested here?

**Comment by brettinternet** [comment]
@yiminc Is this possible? Is it documented anywhere or is the only method to do something like running the server on a port like Temporal CLI's implementation?

I would love to see something similar to the way NATS does this by allowing an in-process connection to the server instead of via the loopback:

```
server, err := server.NewServer(opts)
conn, err := nats.Connect(server.ClientURL(), nats.InProcessServer(server))
```

Here's a related forum topic: https://community.temporal.io/t/embed-the-temporal-server-in-a-go-application/12914/3

**Comment by abtinf** [comment]
@brettinternet I went through all the dev server code and figured out the basic config needed to get embedded working. See https://github.com/abtinf/temporal-a-day/blob/main/001-all-in-one-hello/main.go

**Comment by yg7001** [comment]
@abtinf great job!

**Comment by yiminc** [comment]
Temporal already supports SQLite as persistence for both the core database and visibility store, enabling embedded/single-process deployments with only a database dependency.

The easiest way to run Temporal in embedded mode is via the Temporal CLI:
```
temporal server start-dev
```

**Important caveats:**
- SQLite has single-writer constraints, limiting concurrent write scalability
- Durability guarantees are limited compared to MySQL/PostgreSQL/Cassandra
- This mode is suitable for development, testing, and small-scale on-prem deployments, but is **not recommended for production workloads**

For production deployments, we recommend using MySQL, PostgreSQL, or Cassandra as the persistence layer.

Closing as the requested functionality is available.

**Comment by abtinf** [comment]
@yiminc Clearly that is not what is meant.

I strongly dispute closing as “completed”.

**Comment by sourcec0de** [comment]
@yiminc 
I agree with @abtinf unless we have documentation showing we can import temporal as a Go library and start it. 
Similar to how NATs works
https://dev.to/karanpratapsingh/embedding-nats-in-go-19o


This would be especially useful for testing real clients and workers.
The testify mocking interface is excruciatingly clunky.
The test server is a step in the right direction but often causes flaky tests since it relies on downloading the binary, and the CDN isn't always available.

Testing workflows and activities that rely on context propagation is obtuse.

If we could start workers and clients programmatically in testing and register context propagators the same way we do in production, it would go a long way.

**Comment by yiminc** [comment]
@abtinf & @sourcec0de if i understand it, are you asking for a documentation on how to use temporal server as library? If so we can file a separate issue to track that. We do have samples on how to do that [here](https://github.com/temporalio/samples-server) but a documentation would help.

**Comment by sourcec0de** [comment]
@yiminc isnt that literally in @mfateev's original intent for this ticket? That's why we're in opposition of closing this one.

Also that link doesn't seem to reference using it as a library.

**Comment by yiminc** [comment]
https://github.com/temporalio/documentation/pull/4187

**Comment by yiminc** [comment]
[PR](https://github.com/temporalio/documentation/pull/4187) to update documentations merged, it will be available in next deployment.
